c/swig/typedb_driver_csharp.swg

```diff
@@ -0,0 +1,683 @@
+%noexception role_type_is_deleted;
+%noexception role_type_get_relation_type;
+%noexception role_type_delete;
+%noexception role_type_set_label;
+%noexception role_type_get_supertype;
+%noexception rule_set_label;
+%noexception rule_delete;
+%noexception rule_is_deleted;
+%noexception logic_manager_put_rule;
+%noexception logic_manager_get_rule;
+%noexception query_define;
+%noexception query_undefine;
+%noexception query_delete;
+%noexception query_get_aggregate;
+%noexception transaction_commit;
+%noexception transaction_rollback;
+
+%define %promise(Prefix, Type, function_prefix)
+%typemap(csbody) Prefix ## Promise %{
+    private global::System.Runtime.InteropServices.HandleRef swigCPtr;
+    protected bool swigCMemOwn;
+
+    internal $csclassname(global::System.IntPtr cPtr, bool cMemoryOwn)
+    {
+        swigCMemOwn = cMemoryOwn;
+        swigCPtr = new global::System.Runtime.InteropServices.HandleRef(this, cPtr);
+    }
+
+    internal static global::System.Runtime.InteropServices.HandleRef getCPtr($csclassname obj)
+    {
+        return (obj == null)
+            ? new global::System.Runtime.InteropServices.HandleRef(null, global::System.IntPtr.Zero)
+            : obj.swigCPtr;
+    }
+
+    public Type Resolve()
+    {
+        swigCMemOwn = false;
+        return typedb_driver.function_prefix ## _promise_resolve(this);
+    }
+%}
+
+%typemap(csdisposing, methodname="Dispose", methodmodifiers="protected", parameters="bool disposing") Prefix ## Promise %{{
+    lock(this)
+    {
+        if (swigCPtr.Handle != global::System.IntPtr.Zero)
+        {
+            if (swigCMemOwn)
+            {
+                swigCMemOwn = false;
+                Resolve(); // An equivalent for the standard "delete_xxxPromise".
+            }
+
+            swigCPtr = new global::System.Runtime.InteropServices.HandleRef(null, global::System.IntPtr.Zero);
+        }
+    }
+}%}
+%enddef
+
+%promise(Concept, Concept, concept)
+%promise(Rule, Rule, rule)
+%promise(String, string, string)
+%promise(Bool, bool, bool)
+
+/* void promises require special handling */
+%typemap(csbody) VoidPromise %{
+    private global::System.Runtime.InteropServices.HandleRef swigCPtr;
+    protected bool swigCMemOwn;
+
+    internal $csclassname(global::System.IntPtr cPtr, bool cMemoryOwn)
+    {
+        swigCMemOwn = cMemoryOwn;
+        swigCPtr = new global::System.Runtime.InteropServices.HandleRef(this, cPtr);
+    }
+
+    internal static global::System.Runtime.InteropServices.HandleRef getCPtr($csclassname obj)
+    {
+        return (obj == null)
+            ? new global::System.Runtime.InteropServices.HandleRef(null, global::System.IntPtr.Zero)
+            : obj.swigCPtr;
+    }
+
+    public void Resolve()
+    {
+        swigCMemOwn = false;
+        typedb_driver.void_promise_resolve(this);
+    }
+%}
+
+%typemap(csdisposing, methodname="Dispose", methodmodifiers="protected", parameters="bool disposing") VoidPromise %{{
+    lock(this)
```

Comment:
`Dispose` is auto generated by SWIG and is a method of freeing unmanaged resources when leaving `using` blocks with allocated memory for [`IDisposable` interface](https://learn.microsoft.com/en-us/dotnet/api/system.idisposable?view=net-8.0). You see it here as I use `%typemap(csbody)` typemap to rewrite the body of the generated class, thus erasing this automatically generated method, so I just return it here. 

But yes, they want to prevent race conditions here (with a bit dirty way, tbh). 

On the other hand, `Released` is a method for manual release of our resources in specific situations, similar to Java Driver.